File: samples/FsBulletML.Sample/BaseBullet.fs
namespace FsBulletML.Sample

open System
open System.Collections.Generic 
open Microsoft.Xna.Framework
open Microsoft.FSharp.Core.Operators.Unchecked
open FsBulletML

type BaseBullet () as this =
  [<DefaultValue>]val mutable pos : Vector2
  [<DefaultValue>]val mutable private self : IBullet
  
  interface IBullet with 
    member this.Pos with get () = this.pos 
                     and set (v) = this.pos <- v  
    member this.X with get () = this.pos.X 
                    and set (v) = this.pos.X <- v  
    member this.Y with get () = this.pos.Y
                    and set (v) = this.pos.Y <- v  
    member val AccelerationX = 0.f with get, set
    member val AccelerationY = 0.f with get, set
    member val Speed = 0.f with get, set
    member val Dir = 0.f with get, set
    member val Used = false with get, set
    member val IsBullet = false with get, set
    member val BulletRoot = false with get, set
    member val BulletType = BulletType.Enemy with get, set
    member val ShootingDirection = ShootingDirection.BulletHorizontal with get, set
    member val Task = None with get, set
    member val TargetEnemy = defaultof<IBullet> with get, set
    member val Radius = 0.f with get, set
    member this.Vanish () = this.self.Used <- false
    member this.GetNewBullet() = 
      this.self.BulletRoot <- true

      let newBullet = new BaseBullet() :> IBullet
      newBullet.IsBullet <- true
      newBullet.Radius <- 4.5f
      newBullet.BulletType <- this.self.BulletType
      match newBullet.BulletType with
      | Player -> Manager.addPlayerBullet(newBullet)
      | Enemy -> Manager.addEnemyBullet(newBullet)
      newBullet :> IBulletmlObject

    member this.GetAimDir () : float32 =
      let dir = Math.Atan2( float (BulletMLManager.GetPlayerPosX() - this.self.X),float -(BulletMLManager.GetPlayerPosY() - this.self.Y))
      float32 dir

    member this.GetEnemyAimDir() : float32 = 
      let mutable md = Single.MaxValue 
      if this.self.TargetEnemy :> obj <> null then
        let dir = Math.Atan2( float (this.self.TargetEnemy.X - this.self.X), -1. * float (this.self.TargetEnemy.Y - this.self.Y))
        float32 dir
      else
        for enemy in Manager.enemys do
          let d = Vector2.Distance (Vector2(this.self.X, this.self.Y), Vector2(enemy.X, enemy.Y))
          if md > d then
            this.self.TargetEnemy <- enemy
            md <- d
        let dir = Math.Atan2( float (this.self.TargetEnemy.X - this.self.X), -1. * float (this.self.TargetEnemy.Y - this.self.Y))
        float32 dir

    member this.Init () = 
      this.self.Used <- true
      this.self.BulletRoot <- false

    member this.Update () = this.RunTask()

  member this.RunTask() =
    let apply x y = this.self.X <- x; this.self.Y <- y

    match this.self.Task with
    | None -> ()
    | Some task -> 
      match BulletRunner.run this with
      | true,(x,y) -> 
        apply x y
        this.self.Task |> Option.iter (fun x -> x.Init())
      | false,(x,y) -> apply x y

    if (this.pos.X < 0.f || this.pos.X > Settings.Display.Width || this.pos.Y < 0.f || this.pos.Y > Settings.Display.Height) then
      this.self.Used <- false

  do
    this.self <- this :> IBullet
    this.self.Init()

File: samples/FsBulletML.Sample.Unity2D.CSharp/Assets/Scripts/BaseBullet.cs
using UnityEngine;
using System;
using System.Linq;
using System.Collections.Generic;
using System.Diagnostics;
using FsBulletML;
using Microsoft.FSharp.Core;

public abstract class BaseBullet : MonoBehaviour, FsBulletML.Processable.IBulletmlObject
{
    [SerializeField]
    protected GameObject bulletObject;
    [SerializeField]
    public bool Root { get; set; }
    private GameObject TargetEnemy;
    public abstract GameObject GetBulletPrefubInstance();

    public BaseBullet() : base() {}

    public virtual void Update()
    {
        RunTask();
    }

    protected void RunTask()
    {
        var self = this as FsBulletML.Processable.IBulletmlObject;

        Monad.OptionExtentions.Action<Processable.BulletmlTask>(self.Task,
            x =>
            {
                var result = BulletRunner.Run(this);
                self.X = self.X + (result.X / 100);
                self.Y = self.Y - (result.Y / 100);
                var angle = -(self.Dir) * Mathf.Rad2Deg;
                this.transform.rotation = Quaternion.AngleAxis(angle, new Vector3(0, 0, 1));

                if (result.Processed)
                {
                    x.Init();
                }
                return;
            },
            _ =>
            {
                return;
            });
    }

    public float X
    {
        get { return transform.position.x; }
        set
        {
            var newPosition = this.transform.position;
            newPosition.x = value;
            this.transform.position = newPosition;
        }
    }
    public float Y
    {
        get { return transform.position.y; }
        set
        {
            var newPosition = this.transform.position;
            newPosition.y = value;
            this.transform.position = newPosition;
        }
    }

    public virtual void Init()
    {
        this.Root = false;
        this.Used = true;
        this.BulletRoot = false;
        this.AccelerationX = 0;
        this.AccelerationY = 0;
        this.Speed = 0;
        this.Dir = 0;
        this.IsBullet = true;

        Monad.OptionExtentions.Action<Processable.BulletmlTask>(this.Task,
            x =>
            {
                x.Init();
                return;
            },
            _ =>
            {
                return;
            });

    }

    public void Vanish()
    {
        this.Used = false;
    }

    public float GetAimDir()
    {
        return Mathf.Atan2((Processable.BulletMLManager.GetPlayerPosX() - this.X), (Processable.BulletMLManager.GetPlayerPosY() - this.Y));
    }

    public float GetEnemyAimDir()
    {
        if (this.TargetEnemy != null)
        {
            return Mathf.Atan2(this.TargetEnemy.transform.position.x - this.X, this.TargetEnemy.transform.position.y - this.Y);
        }
        else
        {
            var enemies = GameObject.FindGameObjectsWithTag("Enemy");
            if (!enemies.Any())
            {
                return 0;
            }
            else
            {
                var nearEnemy = enemies.Select(enemy => new { Enemy = enemy, Distance = Vector2.Distance(this.transform.position, enemy.transform.position) }).Min();
                this.TargetEnemy = nearEnemy.Enemy;
                return Mathf.Atan2(this.TargetEnemy.transform.position.x - this.X, this.TargetEnemy.transform.position.y - this.Y);
            }
        }
    }

    public Processable.IBulletmlObject GetNewBullet()
    {
        this.BulletRoot = true;
        var bullet = this.GetBulletPrefubInstance();
        return bullet.GetComponent(typeof(Processable.IBulletmlObject)) as Processable.IBulletmlObject;
    }

    public float AccelerationX { get; set; }
    public float AccelerationY { get; set; }
    public bool BulletRoot { get; set; }
    public Processable.BulletType BulletType { get; set; }
    public float Dir { get; set; }
    public bool IsBullet { get; set; }
    public DTD.ShootingDirection ShootingDirection { get; set; }
    public float Speed { get; set; }
    public Microsoft.FSharp.Core.FSharpOption<Processable.BulletmlTask> Task { get; set; }
    public bool Used { get; set; }
}
